### PR TITLE
[Indexer] bcs deserializer utils

### PR DESCRIPTION
**Pull request by areshand**
### Description

Currently, data in index are still in byte array that is impossible to understand. Provide these functions to deserialize and serialize them back to string correctly.

Demonstrate  deserialization for 
1. byte array in b"" from move
2. bcs serialized byte array
3. property map

### Test Plan
added new tests

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4508)

### CODE REVIEW

File: crates/indexer/src/models/token_bcs_utils.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+use serde::{Serialize, Deserialize, Deserializer};
+use serde::de::Error;
+
+pub fn deserialize_bcs_from_string<'de, D>(deserializer: D) -> Result<String, D::Error>
+    where
+        D: Deserializer<'de>,
+{
+    // convert string to bytes
+    let s = <String>::deserialize(deserializer)?;
+    let decoded = hex::decode(s).expect("Hex string to bcs decoding failed");
```
Comment (bowenyang007) [inline crates/indexer/src/models/token_bcs_utils.rs:10]: Let's propagate error:
```
let decoded = hex::decode(s).map_err(D::Error::custom)?;
```

Comment (areshand) [reply crates/indexer/src/models/token_bcs_utils.rs]: If deserialization fails, will return the original string back to not stop the deserialization. You can also add your customized log in this function

**Comment by bowenyang007** [comment]
I haven't tested property map in action yet but can confirm that deserialize_string_from_hexstring works! Well done sir.

File: crates/indexer/src/models/property_map.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+pub struct PropertyMap {
+    data: HashMap<String, PropertyValue>,
+}
+
+impl PropertyMap {
+    pub fn from_bsc_encode_str(val: Value) -> Option<PropertyMap> {
+        let mut pm = PropertyMap {
+            data: HashMap::new(),
+        };
+        let records: &Vec<Value> = val["map"]["data"].as_array()?;
```
Comment (bowenyang007) [inline crates/indexer/src/models/property_map.rs:31]: Records can be null.

Comment (bowenyang007) [reply crates/indexer/src/models/property_map.rs:31]: Or empty I mean. I don't think that we can assume map or data exists.

File: crates/indexer/src/models/token_bcs_utils.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+{
+    let s = <String>::deserialize(deserializer)?;
+    Ok(convert_bcs_hex("0x1::string::String".to_string(), s.clone()).unwrap_or(s))
+}
+
+
+/// convert the bcs encoded inner value of property_map to its original value in string format
+pub fn deserialize_property_map_from_bcs_hexstring<'de, D>(
+    deserializer: D,
+) -> core::result::Result<String, D::Error>
```
Comment (bowenyang007) [inline crates/indexer/src/models/token_bcs_utils.rs:64]: Let's have this return Value.

File: crates/indexer/src/models/property_map.rs (hunk 2 above)
Comment (bowenyang007) [reply crates/indexer/src/models/property_map.rs:31]: Actually nvm this somehow works. 

I am getting back {"data": {}} though when there's no property. Can we directly return what's inside data?

File: crates/indexer/src/models/token_bcs_utils.rs (hunk 3 above)
Comment (bowenyang007) [reply crates/indexer/src/models/token_bcs_utils.rs:64]:
```
    let str = convert_bcs_propertymap(s.clone()).unwrap_or_else(|| s.to_string());
    serde_json::from_str(&str).map_err(serde::de::Error::custom)
```

File: crates/indexer/src/models/property_map.rs (hunk 2 above)
Comment (bowenyang007) [reply crates/indexer/src/models/property_map.rs:31]: Also tested and it doesn't work as intended. It's always returning {"data": {}}, e.g. when testing with `{“map”: {“data”: [{“key”: “mint_number”, “value”: {“type”: “number”, “value”: “0x30783839”}}]}}`

**Review by bowenyang007 — CHANGES_REQUESTED** [review]
property map still doesn't work. Can we maybe split this PR and ship the single string field for now?

File: crates/indexer/src/models/property_map.rs (hunk 2 above)
Comment (areshand) [reply crates/indexer/src/models/property_map.rs:31]: Quotation mark in your string is “, while the actually quotation mark should be ". If you replace the quotation mark with the correct one, it should work.

File: crates/indexer/src/models/token_bcs_utils.rs (hunk 3 above)
Comment (areshand) [reply crates/indexer/src/models/token_bcs_utils.rs]: updated to return value instead of string

**Comment by davidiw** [comment]
closing out stale pr

**Comment by bowenyang007** [comment]
wait i need this. let me work on it.

**Comment by bowenyang007** [comment]
Oh nvm I can't work on PR. I'll just copy the changes over.

**Pull request closed without merging**